# Census Data Cleaning - by Felix Emmanuel 202122699

### Reading in our csv file

In [1]:
import pandas as pd
import seaborn as sns
import matplotlib.pyplot as plt

In [2]:
df = pd.read_csv("census_20.csv")

   ### Explorative Data Analysis (EDA)

In [3]:
df.info()


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10610 entries, 0 to 10609
Data columns (total 11 columns):
 #   Column                         Non-Null Count  Dtype 
---  ------                         --------------  ----- 
 0   House Number                   10610 non-null  int64 
 1   Street                         10610 non-null  object
 2   First Name                     10610 non-null  object
 3   Surname                        10610 non-null  object
 4   Age                            10610 non-null  object
 5   Relationship to Head of House  10610 non-null  object
 6   Marital Status                 7995 non-null   object
 7   Gender                         10610 non-null  object
 8   Occupation                     10610 non-null  object
 9   Infirmity                      10610 non-null  object
 10  Religion                       7928 non-null   object
dtypes: int64(1), object(10)
memory usage: 911.9+ KB


##### From the info we could see that we have 10610 rows and 11 columns 
##### I could deduce that something is wrong with "Age" series because the 'Dtype' should be an  integer and not object
#### I could also spot that "Marital Status" and "Religion" columns has null values

In [4]:
df.shape

(10610, 11)

In [5]:
print(df.to_string()) 
#a view of the entire data frame, however this is not so organised to explore


       House Number               Street   First Name                 Surname                Age Relationship to Head of House Marital Status  Gender                                                           Occupation            Infirmity   Religion
0                 1         Smith Radial        Grace                   Patel                 46                          Head        Widowed  Female                                                   Petroleum engineer                 None   Catholic
1                 1         Smith Radial          Ian                   Nixon                 24                        Lodger         Single    Male                                            Publishing rights manager                 None  Christian
2                 2         Smith Radial    Frederick                    Read                 87                          Head       Divorced    Male                                                 Retired TEFL teacher                 None   Catho

In [6]:
df

House Number        Street First Name   Surname Age  \
0                 1  Smith Radial      Grace     Patel  46   
1                 1  Smith Radial        Ian     Nixon  24   
2                 2  Smith Radial  Frederick      Read  87   
3                 3  Smith Radial     Daniel     Adams  58   
4                 3  Smith Radial    Matthew      Hall  13   
...             ...           ...        ...       ...  ..   
10605             1     Page Farm       Leon   Gregory  42   
10606             1     Page Farm      Sally   Gregory   4   
10607             1     Page Farm     Graham   Gregory   1   
10608             1     Page Farm       Toby  Jennings  31   
10609             1     Page Farm    Cameron     Patel  28   

      Relationship to Head of House Marital Status  Gender  \
0                              Head        Widowed  Female   
1                            Lodger         Single    Male   
2                              Head       Divorced    Male   
3                              Head       Divorced    Male   
4                          Grandson            NaN    Male   
...                             ...            ...     ...   
10605                       Husband        Married    Male   
10606                      Daughter            NaN  Female   
10607                           Son            NaN    Male   
10608                        Lodger         Single    Male   
10609                        Lodger         Single    Male   

                        Occupation Infirmity   Religion  
0               Petroleum engineer      None   Catholic  
1        Publishing rights manager      None  Christian  
2             Retired TEFL teacher      None   Catholic  
3                 Therapist, music      None   Catholic  
4                          Student      None        NaN  
...                            ...       ...        ...  
10605  Passenger transport manager      None       None  
10606                        Child      None        NaN  
10607                        Child      None        NaN  
10608                   Unemployed      None  Christian  
10609         Commissioning editor      None       None  

[10610 rows x 11 columns]

In [7]:
df.describe()

House Number
count  10610.000000
mean      40.145994
std       40.393975
min        1.000000
25%       11.000000
50%       27.000000
75%       54.000000
max      228.000000

## Data Munging

### Next I will call out the unique value of each column for inspection and error detection

In [8]:
df["House Number"].unique()


array([  1,   2,   3,   4,   5,   6,   7,   8,   9,  10,  11,  12,  13,
        14,  15,  16,  17,  18,  19,  20,  21,  22,  23,  24,  25,  26,
        27,  28,  29,  30,  31,  32,  33,  34,  35,  36,  37,  38,  39,
        40,  41,  42,  43,  44,  45,  46,  47,  48,  49,  50,  51,  52,
        53,  54,  55,  56,  57,  58,  59,  60,  61,  62,  63,  64,  65,
        66,  67,  68,  69,  70,  71,  72,  73,  74,  75,  76,  77,  78,
        79,  80,  81,  82,  83,  84,  85,  86,  87,  88,  89,  90,  91,
        92,  93,  94,  95,  96,  97,  98,  99, 100, 101, 102, 103, 104,
       105, 106, 107, 108, 109, 110, 111, 112, 113, 114, 115, 116, 117,
       118, 119, 120, 121, 122, 123, 124, 125, 126, 127, 128, 129, 130,
       131, 132, 133, 134, 135, 136, 137, 138, 139, 140, 141, 142, 143,
       144, 145, 146, 147, 148, 149, 150, 151, 152, 153, 154, 155, 156,
       157, 158, 159, 160, 161, 162, 163, 164, 165, 166, 167, 168, 169,
       170, 171, 172, 173, 174, 175, 176, 177, 178, 179, 180, 18

##### ERRORS :  None found, array is good and all unique values are integers, no float, no empty space!

In [9]:
df["Street"].unique()

array(['Smith Radial', 'Hodgson Shoals', 'January Avenue', 'Green Meadow',
       'Richards Locks', 'Knight Road', 'Stoke Port', 'Carlisle Meadows',
       'Laurel Common', 'Turkey Cliff', 'Gough Mission',
       'Breakfast Drive', 'Crab Orchard', 'Giles Avenue',
       'Copper Overpass', 'Young Street', 'Kraken Drive',
       'Athollbird Avenue', 'Warren Trafficway', 'Beverley Estate',
       'Hussain Street', 'Orchard Forges', 'Conifer Walks',
       'Clearwater Drive', 'Mustard Roads', 'Elvis Gardens',
       'Brown Avenue', 'Preston Avenue', 'Brown Villages',
       'Dockers Plains', 'Zigzag Brooks', 'Ouse Road', 'Beaver Port',
       'Lichfieldnet Street', 'Power Landing', 'Barnett Lane',
       'Hudson Gateway', 'Thames Forges', 'Gwent Road', 'Cox Mills',
       'Phillips Drive', 'Powell Drive', 'Birmingham Spurs',
       'Mill Avenue', 'Harrison Drive', 'Henry Underpass',
       'Wolverinetun Drive', 'Foxglove Street', 'Thompson Fields',
       'George Ferry', 'Goldfinch Drive',

##### ERRORS : No errors found. No null values, street names look ok by inspection, all names started with Upper case letters



In [10]:
df["First Name"].unique()

array(['Grace', 'Ian', 'Frederick', 'Daniel', 'Matthew', 'Steven',
       'Alison', 'Kelly', 'Kim', 'Oliver', 'Jemma', 'Terence', 'Mohamed',
       'Patricia', 'Kenneth', 'Gerald', 'Elliot', 'Adrian', 'Hilary',
       'Lynda', 'Eric', 'Joe', 'Ross', 'Rebecca', 'Sophie', 'Aaron',
       'Elizabeth', 'Graeme', 'Jordan', 'Amy', 'Janice', 'Hugh', 'Paula',
       'Martyn', 'Sharon', 'Cheryl', 'Ben', 'Vanessa', 'Arthur', 'Sara',
       'Kevin', 'Paige', 'Darren', 'Kerry', 'Kimberley', 'Damian',
       'Patrick', 'June', 'Gillian', 'Josephine', 'Keith', 'Brett',
       'Carly', 'Duncan', 'Rachel', 'Shaun', 'Maurice', 'Denis',
       'Samantha', 'Lewis', 'Abbie', 'Leslie', 'Marian', 'Antony',
       'Raymond', 'Diana', 'Robert', 'Barbara', 'Brandon', 'Bethan',
       'Sean', 'Irene', 'Russell', 'Jill', 'Tom', 'Robin', 'Leonard',
       'Joan', 'Jacob', 'Edward', 'Josh', 'Nicola', 'Thomas', 'Bernard',
       'Lee', 'Sandra', 'Naomi', 'Marcus', 'Hollie', 'Natalie', 'Pauline',
       'Stuart', 'R

##### ERRORS : No errors found in the "First Name". No null values, names look ok by inspection, all names started with upper case letters, no wrong spellings also.



In [11]:
df["Surname"].unique()

array(['Patel', 'Nixon', 'Read', 'Adams', 'Hall', 'Fletcher', 'Jenkins',
       'Browne', 'Hussain', 'Jones', 'Wilson', 'Heath', 'Ellis', 'Day',
       'Hughes', 'Marsh', 'Gregory', 'Bailey', 'Stephens', 'Rogers',
       'Thomas', 'Banks', 'Reid', 'McDonald', 'Richards', 'Hopkins',
       'Shaw', 'Dunn', 'Ashton', 'Khan', 'Jackson', 'Thompson',
       'McCarthy', 'Clarke', "O'Neill", 'Williams', 'Hawkins', 'Kemp',
       'Chadwick', 'Webster', 'Johnson', 'Hudson', 'Smith', 'Roberts',
       'Young', 'Moss', 'Matthews-Moss', 'Evans', 'Green', 'Brown',
       'Newman', 'Black', 'Lamb', 'Payne', 'James', 'Townsend', 'Barnes',
       'Gibbons', 'Cook', 'Kelly', 'Humphries', 'Scott', 'Hill', 'Watson',
       'Watson-Williams', 'Hancock', 'Goodwin', 'Stevens', 'Davies',
       'Cox', 'Hooper', 'Burrows', 'George', 'Noble', 'Baker', 'Mills',
       'Wright', 'Powell', 'Lewis', 'Norman', 'Bray', 'Murray', 'Walsh',
       'Hunt', 'Summers', 'Dyer', 'Waters', 'Buckley', 'Walton', 'Lee',
       '

##### ERRORS : No errors found in the 'Surname' series. No null values, names look intact by inspection, all names started with upper cases.



### AGE

In [12]:
df["Age"].unique()

array(['46', '24', '87', '58', '13', '9', '38', '12', '69', '42', '48',
       '8', '33', '16', '6', '4', '74', '70', '40', '82', '28', '71',
       '72', '86', '84', '34', '5', '2', '47', '20', '39', '23', '30',
       '31', '1', '60', '105', '102', '43', '41', '10', '50', '14', '49',
       '11', '78', '37', '45', '35', '61', '18', '52', '51', '15', '7',
       '65', '27', '73', '59', '54', '17', '36', '63', '3', '62', '21',
       '0', '32', '29', '83', '68', '53', '67', '80', '88', '100', '76',
       '22', '85', '75', '81', '26', '66', '19', '94', '55', '64', '56',
       '44', '57', '25', '89', '77', '79', '97', '95', '98', '104',
       '43.54302670766108', '49.54302670766108', '5.0', '103', ' ', '91',
       '90', '96', '89.33008765095417', '94.33008765095417', '41.0',
       '38.0', '99', '93', '89.2477665064506', '112'], dtype=object)

##### ERRORS : There are errors in the "Age" series that need to be fixed. Floats numbers are not correct representation of Age, Empty cells need to be fixed, outliiers need to be questioned and inspected.


## Inspecting, analyzing and fixing the empty cells in the 'Age' Series.

In [13]:
print(df[df['Age']==' '].index.values) 

[2480 4575 5885 6746 9900]


In [14]:
df.loc[2480]


House Number                                27
Street                           Mustard Roads
First Name                           Stephanie
Surname                                  Brown
Age                                           
Relationship to Head of House         Daughter
Marital Status                             NaN
Gender                                  Female
Occupation                             Student
Infirmity                                 None
Religion                                   NaN
Name: 2480, dtype: object

In [15]:
df.loc[2478 :2482]

House Number         Street First Name   Surname Age  \
2478            27  Mustard Roads    William  Lawrence  40   
2479            27  Mustard Roads      Donna     Brown  17   
2480            27  Mustard Roads  Stephanie     Brown       
2481            27  Mustard Roads      Susan     Brown  13   
2482            28  Mustard Roads    Deborah    Knight  27   

     Relationship to Head of House Marital Status  Gender         Occupation  \
2478                       Partner         Single    Male  Therapist, sports   
2479                      Daughter            NaN  Female            Student   
2480                      Daughter            NaN  Female            Student   
2481                      Daughter            NaN  Female            Student   
2482                          Head        Married  Female    Event organiser   

     Infirmity   Religion  
2478      None       None  
2479      None        NaN  
2480      None        NaN  
2481      None        NaN  
2482      None  Christian

In [16]:
#Inspecting the entries, index 2480 is one of the three daughters of the househould with ages of 17 and 13.
#they are all students so I took average of the two other ages
#it's suffice to say the average is a comfortable age bracket, plus it makes the ages a good sequence of 17, 15, and 13
#a possible case of a planned decision from the parents to give 2 years interval each between each birth

In [17]:
#fixing the missing cell 
df.loc[2480, 'Age'] = 15

In [18]:
df.loc[4575]

House Number                                 96
Street                           Hudson Gateway
First Name                              Jasmine
Surname                                  Gibson
Age                                            
Relationship to Head of House              Head
Marital Status                           Single
Gender                                   Female
Occupation                          Illustrator
Infirmity                                  None
Religion                               Catholic
Name: 4575, dtype: object

In [19]:
df.loc[4573:4577]

House Number          Street First Name  Surname Age  \
4573            95  Hudson Gateway   Michelle  Simpson  26   
4574            95  Hudson Gateway      Carly  Simpson  21   
4575            96  Hudson Gateway    Jasmine   Gibson       
4576            96  Hudson Gateway      Kelly   Gibson  10   
4577            97  Hudson Gateway     Kirsty    Jones  45   

     Relationship to Head of House Marital Status  Gender  \
4573                      Daughter         Single  Female   
4574                      Daughter         Single  Female   
4575                          Head         Single  Female   
4576                 Granddaughter            NaN  Female   
4577                          Head         Single  Female   

                  Occupation Infirmity   Religion  
4573  Operational researcher      None  Methodist  
4574      University Student      None       None  
4575             Illustrator      None   Catholic  
4576                 Student      None        NaN  
4577              Orthoptist      None   Catholic

In [20]:
#Calling out rows above and beneath the missing "Age" cell, we could in index 4575 that
#Jasmine has a 10 years old granddaughter living with her, so this is a factor I considered in giving her an age of 60.
#Also, going throug the row
#her marital status is a false record because she has a grand daughter, hence it is impossible for her to be single
#either she is married, widowed or divorced but since the husband wasn't captured, it is better to assume she is widowed

In [21]:
#fixing the age and marital status simultaneously
df.loc[4575, 'Age'] = 60
df.loc[4575, 'Marital Staus'] = "Widowed"

In [22]:
df.loc[5885]

House Number                               12
Street                           Henge Street
First Name                               John
Surname                                Harris
Age                                          
Relationship to Head of House             Son
Marital Status                            NaN
Gender                                   Male
Occupation                            Student
Infirmity                                None
Religion                                  NaN
Marital Staus                             NaN
Name: 5885, dtype: object

In [23]:
df.loc[5883:5887]

House Number        Street First Name Surname Age  \
5883            11  Henge Street       Tina   Green  40   
5884            12  Henge Street      Aimee  Harris  57   
5885            12  Henge Street       John  Harris       
5886            13  Henge Street     Hilary  Nelson  25   
5887            13  Henge Street     Norman  Nelson   7   

     Relationship to Head of House Marital Status  Gender    Occupation  \
5883                       Partner         Single  Female  Risk analyst   
5884                          Head       Divorced  Female    Unemployed   
5885                           Son            NaN    Male       Student   
5886                          Head       Divorced  Female    Unemployed   
5887                           Son            NaN    Male       Student   

     Infirmity   Religion Marital Staus  
5883      None   Catholic           NaN  
5884      None  Christian           NaN  
5885      None        NaN           NaN  
5886      None   Catholic           NaN  
5887      None        NaN           NaN

In [24]:
#Deciding the age of John Harris (5885) is quite a tough one, as student age ranges from about 3 years to about 35 for mature students. 
#However since he's still living with his parent, I will assume he is not a mature student,
#Therefore , I will take the range of this age and then half it. i.e
#((35-3)/2) = 16

df.loc[5885, 'Age'] = 16

In [25]:
df.loc[6746]

House Number                               44
Street                           South Avenue
First Name                             Arthur
Surname                              Mitchell
Age                                          
Relationship to Head of House             Son
Marital Status                            NaN
Gender                                   Male
Occupation                            Student
Infirmity                                None
Religion                                  NaN
Marital Staus                             NaN
Name: 6746, dtype: object

In [26]:
df.loc[6744:6748]

House Number        Street First Name   Surname Age  \
6744            44  South Avenue        Max  Mitchell  42   
6745            44  South Avenue     Claire   Parsons  40   
6746            44  South Avenue     Arthur  Mitchell       
6747            45  South Avenue       Eric    Hewitt  71   
6748            45  South Avenue      Naomi    Hewitt  68   

     Relationship to Head of House Marital Status  Gender  \
6744                          Head         Single    Male   
6745                       Partner         Single  Female   
6746                           Son            NaN    Male   
6747                          Head        Married    Male   
6748                          Wife        Married  Female   

                                  Occupation Infirmity   Religion  \
6744                         Games developer      None       None   
6745      Government social research officer      None       None   
6746                                 Student      None        NaN   
6747  Retired Manufacturing systems engineer      None  Christian   
6748           Retired Environmental manager      None       None   

     Marital Staus  
6744           NaN  
6745           NaN  
6746           NaN  
6747           NaN  
6748           NaN

In [27]:
#inspecting the rows the only indicator we have is that index 6746 is a student. He is living alone because 
#he is not related to the entries above or beneath. Therefore he is the head of household, which infers
#he's most likely a university student and hence I'll give him a random age of 21, 
#also I will change his relationshp to 'Head' 
#by extension of this age, I will set his 'NaN' marital status to 'Single'

df.loc[6746, 'Age'] = 21
df.loc[6746, 'Relationship to Head of House'] = "Head"
df.loc[6746, 'Marital Status'] = "Single"

In [28]:
df.loc[9900]

House Number                                    61
Street                                Regent Coves
First Name                                   Susan
Surname                                       Ross
Age                                               
Relationship to Head of House             Daughter
Marital Status                            Divorced
Gender                                      Female
Occupation                       Nurse, children's
Infirmity                                     None
Religion                                 Christian
Marital Staus                                  NaN
Name: 9900, dtype: object

In [29]:
df.loc[9898:9903]

House Number        Street First Name    Surname Age  \
9898            60  Regent Coves     Sylvia  O'Donnell  43   
9899            61  Regent Coves    Pauline       Ross  66   
9900            61  Regent Coves      Susan       Ross       
9901            61  Regent Coves     Joanne       Ross  29   
9902            61  Regent Coves      Allan       Ross  29   
9903            62  Regent Coves   Michelle     Hanson  64   

     Relationship to Head of House Marital Status  Gender  \
9898                          Wife        Married  Female   
9899                          Head        Widowed  Female   
9900                      Daughter       Divorced  Female   
9901                      Daughter         Single  Female   
9902                           Son         Single    Male   
9903                          Head        Married  Female   

                  Occupation Infirmity   Religion Marital Staus  
9898    Engineer, biomedical      None       None           NaN  
9899     Immigration officer      None  Christian           NaN  
9900       Nurse, children's      None  Christian           NaN  
9901                Animator      None       None           NaN  
9902              Unemployed      None  Christian           NaN  
9903  Financial risk analyst      None  Christian           NaN

In [30]:
#Inspecting the entries above and below  cell 9900 , we could tell that this person lives with her mother and siblings
#since she's divorced and her siblings (twins) are 29 years already, we know she is in her thirties. 
#Hence, I am most inclined to peg her age at 33.

df.loc[9900, 'Age'] = 33


In [31]:
#lets take a look at our "Age" series now after removing the empty entries
df["Age"].unique()

array(['46', '24', '87', '58', '13', '9', '38', '12', '69', '42', '48',
       '8', '33', '16', '6', '4', '74', '70', '40', '82', '28', '71',
       '72', '86', '84', '34', '5', '2', '47', '20', '39', '23', '30',
       '31', '1', '60', '105', '102', '43', '41', '10', '50', '14', '49',
       '11', '78', '37', '45', '35', '61', '18', '52', '51', '15', '7',
       '65', '27', '73', '59', '54', '17', '36', '63', '3', '62', '21',
       '0', '32', '29', '83', '68', '53', '67', '80', '88', '100', '76',
       '22', '85', '75', '81', '26', '66', '19', '94', '55', '64', '56',
       '44', '57', '25', '89', '77', '79', '97', '95', '98', '104',
       '43.54302670766108', '49.54302670766108', '5.0', '103', 15, '91',
       '90', '96', '89.33008765095417', '94.33008765095417', '41.0',
       '38.0', '99', '93', 60, '89.2477665064506', 16, 21, '112', 33],
      dtype=object)

In [32]:
#cross examining the ages that are questionable, i.e 0 and 112. 
#After checking throgh the ages '0', I discovered that they are children attached to Households, Hence I can say thy are
#kids less than a year old
#112 calls for questioning , if this a real age as this is an outlier.

print(df[df['Age']=='112'].index.values)

[8874]


In [33]:
df.loc[8871:8876]

#the analysis shows this entry is real as he is retired, married to a 102 years old woman (also retired)
#with a son of 32 years. This therefore validates the entry.

House Number        Street First Name  Surname  Age  \
8871            52  Smith Points     Stacey   Talbot   27   
8872            52  Smith Points    Francis  Johnson   32   
8873            53  Smith Points      Amber  Johnson  102   
8874            53  Smith Points      Billy  Johnson  112   
8875            54  Smith Points     Brenda   Parkin   18   
8876            54  Smith Points    Cameron   Willis   20   

     Relationship to Head of House Marital Status  Gender  \
8871                          Head         Single  Female   
8872                          None         Single    Male   
8873                          Head        Married  Female   
8874                       Husband        Married    Male   
8875                          Head         Single  Female   
8876                          None         Single    Male   

                            Occupation Infirmity   Religion Marital Staus  
8871          Advertising art director      None       None           NaN  
8872        Surveyor, building control      None   Catholic           NaN  
8873  Retired Teacher, adult education      None  Christian           NaN  
8874                 Retired Press sub      None       None           NaN  
8875                University Student      None       None           NaN  
8876                University Student      None  Christian           NaN

### Converting the age type to integers

In [36]:
age= df['Age'].astype(float).astype(int)

In [38]:
#to confirm if our series has now changed to integer
age.unique()

array([ 46,  24,  87,  58,  13,   9,  38,  12,  69,  42,  48,   8,  33,
        16,   6,   4,  74,  70,  40,  82,  28,  71,  72,  86,  84,  34,
         5,   2,  47,  20,  39,  23,  30,  31,   1,  60, 105, 102,  43,
        41,  10,  50,  14,  49,  11,  78,  37,  45,  35,  61,  18,  52,
        51,  15,   7,  65,  27,  73,  59,  54,  17,  36,  63,   3,  62,
        21,   0,  32,  29,  83,  68,  53,  67,  80,  88, 100,  76,  22,
        85,  75,  81,  26,  66,  19,  94,  55,  64,  56,  44,  57,  25,
        89,  77,  79,  97,  95,  98, 104, 103,  91,  90,  96,  99,  93,
       112])

## Relationship to Head of House

In [39]:
df["Relationship to Head of House"].unique()

array(['Head', 'Lodger', 'Grandson', 'Daughter', 'Husband', 'None', 'Son',
       'Adopted Son', 'Wife', 'Partner', 'Step-Daughter', 'Step-Son',
       'Granddaughter', 'Cousin', 'Visitor', 'Sibling',
       'Adopted Daughter', 'Son-in-law', 'Neice', 'Nephew',
       'Adopted Granddaughter'], dtype=object)

##### ERRORS : "NONE" entry needs to changed, also "Neice" needs to correctly spelt as "Niece"

####                                          Fixing errors for "Relationship to Head of House" we have:

In [40]:
#it only makes sense to assume that people who filled "None" are not members of the household hence classed as visitors
df['Relationship to Head of House'].replace(["Neice"], 'Niece', inplace=True)
df['Relationship to Head of House'].replace(["None"], 'Visitor', inplace=True)

In [41]:
df["Relationship to Head of House"].unique()

array(['Head', 'Lodger', 'Grandson', 'Daughter', 'Husband', 'Visitor',
       'Son', 'Adopted Son', 'Wife', 'Partner', 'Step-Daughter',
       'Step-Son', 'Granddaughter', 'Cousin', 'Sibling',
       'Adopted Daughter', 'Son-in-law', 'Niece', 'Nephew',
       'Adopted Granddaughter'], dtype=object)

## Marital Status

In [42]:
df["Marital Status"].unique()

array(['Widowed', 'Single', 'Divorced', nan, 'Married'], dtype=object)

###### ERRORS : "nan" entry needs to filled

In [43]:
No_of_rows = df.shape[0]
No_of_columns = df.shape[1]
print("Rows: ", No_of_rows)
print("Columns: ", No_of_columns)

Rows:  10610
Columns:  12


In [44]:
# The Large columns with null values however cannot be dropped as these rows constitute a large part of the dataset 
percentage_of_null_values_marital_status_col = (2615/ No_of_rows)*100
print("% Null MS: ",percentage_of_null_values_marital_status_col)

% Null MS:  24.64655984919887


In [45]:
df["Marital Status"].fillna(method="pad", inplace=True)

In [47]:
df["Marital Status"].isnull().sum()

0

In [48]:
df["Marital Status"].unique()

array(['Widowed', 'Single', 'Divorced', 'Married'], dtype=object)

## Gender

In [49]:
df['Gender'].unique()

array(['Female', 'Male'], dtype=object)

## Occupation

In [50]:
df['Occupation'].unique()

array(['Petroleum engineer', 'Publishing rights manager',
       'Retired TEFL teacher', ...,
       'Retired Scientist, clinical (histocompatibility and immunogenetics)',
       'Retired Theatre manager',
       'Retired Engineer, broadcasting (operations)'], dtype=object)

## Infirmity

In [51]:
df['Infirmity'].unique()

array(['None', ' ', 'Deaf', 'Mental Disability', 'Unknown Infection',
       'Disabled', 'Physical Disability', 'Blind'], dtype=object)

#### The entries contain blank cells, i will call out these cells for inspection

In [52]:
print(df[df['Infirmity']==' '].index.values)

[  34 1726 2193 2924 4369 5894 6646 6710 8083 8811 8982 9508]


In [54]:
df.loc[33:36]

House Number        Street First Name   Surname  Age  \
33            13  Smith Radial      Paula   Gregory    2   
34            13  Smith Radial     Martyn    Bailey   47   
35            13  Smith Radial     Sharon  Stephens   20   
36            14  Smith Radial     Cheryl    Rogers   39   

   Relationship to Head of House Marital Status  Gender          Occupation  \
33                      Daughter         Single  Female               Child   
34                        Lodger       Divorced    Male  Wellsite geologist   
35                        Lodger         Single  Female  University Student   
36                          Head       Divorced  Female   Nurse, children's   

   Infirmity   Religion Marital Staus  
33      None        NaN           NaN  
34                 None           NaN  
35      None  Christian           NaN  
36      None  Christian           NaN

In [55]:
df.loc[1725:1728]

House Number          Street First Name Surname  Age  \
1725             2  Hussain Street    Shannon  Davies   19   
1726             2  Hussain Street      Rosie  Davies   15   
1727             2  Hussain Street      Abdul  Davies   13   
1728             3  Hussain Street     Judith  Hewitt   45   

     Relationship to Head of House Marital Status  Gender          Occupation  \
1725                      Daughter         Single  Female  University Student   
1726                      Daughter         Single  Female             Student   
1727                           Son         Single    Male             Student   
1728                          Head        Married  Female   Transport planner   

     Infirmity Religion Marital Staus  
1725      None     None           NaN  
1726                NaN           NaN  
1727      Deaf      NaN           NaN  
1728      None     None           NaN

In [56]:
df.loc[2191:2194]

House Number         Street First Name   Surname  Age  \
2191            69  Conifer Walks      Julia     Baker   56   
2192            70  Conifer Walks      Allan    Bailey   73   
2193            70  Conifer Walks  Stephanie    Bailey   74   
2194            71  Conifer Walks      Gavin  Richards   81   

     Relationship to Head of House Marital Status  Gender  \
2191                          Wife        Married  Female   
2192                          Head        Married    Male   
2193                          Wife        Married  Female   
2194                          Head        Married    Male   

                                         Occupation          Infirmity  \
2191                       Customer service manager               None   
2192  Retired Product/process development scientist               None   
2193      Retired Environmental health practitioner                      
2194                         Retired Futures trader  Unknown Infection   

       Religion Marital Staus  
2191       None           NaN  
2192  Christian           NaN  
2193  Christian           NaN  
2194  Methodist           NaN

In [58]:
df.loc[2922:2926]

House Number     Street  First Name   Surname  Age  \
2922            47  Ouse Road      Pamela     Jones   59   
2923            48  Ouse Road     Jeffrey  Lawrence   42   
2924            48  Ouse Road     Kathryn  Lawrence   37   
2925            48  Ouse Road       Scott  Lawrence    7   
2926            48  Ouse Road  Jacqueline  Lawrence    7   

     Relationship to Head of House Marital Status  Gender  \
2922                          Head         Single  Female   
2923                          Head        Married    Male   
2924                          Wife        Married  Female   
2925                           Son        Married    Male   
2926                      Daughter        Married  Female   

                      Occupation Infirmity   Religion Marital Staus  
2922  Civil engineer, consulting      None  Methodist           NaN  
2923                 Ship broker      None       None           NaN  
2924      Psychologist, clinical                 None           NaN  
2925                     Student      None        NaN           NaN  
2926                     Student      None        NaN           NaN

In [59]:
#since we cannot transfer infirmity as an attribute, and we have quite a small number of missing entries, 
#it's best to replace these entries with 'None'
df['Infirmity'].replace(' ', 'None', inplace = True)


## Religion

In [60]:
df["Religion"].unique()

array(['Catholic', 'Christian', nan, 'None', 'Methodist', 'Sikh',
       'Muslim', 'Undecided', 'Jewish', 'Bahai', ' ', 'Pagan', 'Private',
       'Buddist', 'Sith', 'Quaker', 'Agnostic', 'Nope'], dtype=object)

#### ERRORS: We have 'nan', blank cells, 'Nope' is not a religion, this will be replaced with 'None'

In [61]:
No_of_rows = df.shape[0]
No_of_columns = df.shape[1]
print("Rows: ", No_of_rows)
print("Columns: ", No_of_columns)

Rows:  10610
Columns:  12


In [62]:
# The Large columns with null values however cannot be dropped as these rows constitute a large part of the dataset 
percentage_of_null_values_religion_col = (2682/ No_of_rows)*100
print("% Null Religion: ",percentage_of_null_values_religion_col)

% Null Religion:  25.27803958529689


In [63]:
print(df[df['Religion']==' '].index.values)

[ 2168  4891 10495]


In [64]:
print(df[df['Religion']=='Private'].index.values)

[ 6512  6513 10274 10275]


In [66]:
df.loc[2166:2170]

House Number         Street  First Name   Surname  Age  \
2166            60  Conifer Walks        Rhys     Smith   14   
2167            61  Conifer Walks  Jacqueline  Mitchell   26   
2168            61  Conifer Walks       Garry   Brennan   28   
2169            61  Conifer Walks       Bruce  Mitchell    9   
2170            62  Conifer Walks   Christine   Russell   22   

     Relationship to Head of House Marital Status  Gender  \
2166                      Step-Son         Single    Male   
2167                          Head         Single  Female   
2168                       Partner         Single    Male   
2169                           Son         Single    Male   
2170                          Head        Married  Female   

                    Occupation Infirmity  Religion Marital Staus  
2166                   Student      None       NaN           NaN  
2167  Administrator, education      None      Sikh           NaN  
2168    Health service manager      None                     NaN  
2169                   Student      None       NaN           NaN  
2170        University Student      None  Catholic           NaN

In [67]:
df.loc[4889:4893]
#religion is not a transferable attribute by association or relationship,
#hence, this missing values will be replaced with "None"

House Number            Street First Name   Surname  Age  \
4889            21  Birmingham Spurs   Patricia  Mitchell   25   
4890            21  Birmingham Spurs       John  Mitchell   25   
4891            22  Birmingham Spurs      Paige   Chapman   27   
4892            22  Birmingham Spurs      Robin   Chapman   26   
4893            22  Birmingham Spurs     Howard   Chapman   10   

     Relationship to Head of House Marital Status  Gender  \
4889                          Head        Married  Female   
4890                       Husband        Married    Male   
4891                          Head        Married  Female   
4892                       Husband        Married    Male   
4893                           Son        Married    Male   

                 Occupation Infirmity   Religion Marital Staus  
4889             Unemployed      None  Christian           NaN  
4890          Hotel manager      None       None           NaN  
4891  Editor, commissioning      None                      NaN  
4892         Engineer, land      None       None           NaN  
4893                Student      None        NaN           NaN

In [68]:
df.loc[10493:10497]

House Number         Street First Name    Surname  Age  \
10493             1  Booth Circles    Matthew       Owen   70   
10494             1  Booth Circles      Paige       Owen   68   
10495             2  Booth Circles   Samantha  Pritchard   54   
10496             2  Booth Circles       Jake  Pritchard   56   
10497             2  Booth Circles     Stuart  Pritchard   16   

      Relationship to Head of House Marital Status  Gender  \
10493                          Head        Married    Male   
10494                          Wife        Married  Female   
10495                          Head        Married  Female   
10496                       Husband        Married    Male   
10497                           Son        Married    Male   

                                Occupation Infirmity   Religion Marital Staus  
10493  Retired Information systems manager      None   Catholic           NaN  
10494        Retired Multimedia programmer      None   Catholic           NaN  
10495                      Learning mentor      None                      NaN  
10496                  Designer, furniture      None  Christian           NaN  
10497                              Student      None        NaN           NaN

In [69]:
df['Religion'].replace(' ', 'None', inplace = True)



In [70]:
#removing Nan
df["Religion"].fillna(method="pad", inplace=True)

In [71]:
df["Religion"].isnull().sum()

0

In [72]:
#replacing "Nope" 
df['Religion'].replace('Nope', 'None', inplace = True)

In [73]:
#the religion types have all been researched to be valid
df['Religion'].unique()

array(['Catholic', 'Christian', 'None', 'Methodist', 'Sikh', 'Muslim',
       'Undecided', 'Jewish', 'Bahai', 'Pagan', 'Private', 'Buddist',
       'Sith', 'Quaker', 'Agnostic'], dtype=object)

### SAVING OUR CLEANED DATA (for data visualization and analysis in another Jupyter notebook)

In [74]:
df.to_csv("cleaned census_20.csv")